### ISSUE TITLE
Automatically adjust horizontal boxes in table of contents

### ISSUE THREAD

**Issue by dbitouze** [latex3/latex2e#1272]
## Brief outline of the enhancement

Automatically adjust horizontal boxes in table of contents.

## Minimal example showing the current behaviour

It would be nice not to have to manually adjust the width of horizontal boxes (for example with `\renewcommand{\@pnumwidth}{...}`) in cases like:

```latex
\RequirePackage{latexbug}       % <--should be always the first line (see CONTRIBUTING)!
\documentclass[draft]{article}
\begin{document}
\tableofcontents
\setcounter{page}{1000}
\setcounter{section}{100}
\section{Foo}
\subsection{Foooo}
\end{document}
```
![image](https://github.com/latex3/latex2e/assets/1032633/8a41a334-eff8-4863-8048-bd72e75ab220)

**Comment by davidcarlisle** [comment]
You mean in case you spend 30 years enlarging `source2e.tex` and accidentally go over a thousand pages:-)

This has long been known as a common issue with larger documents (or documents with unusual page formatting) It wouldn't be particularly difficult to keep track of the width and store the maximum (or less computationally expensive just keep track of the largest page (and section) number and assume the typeset version is monotonically increasing in width) you could then write them to the aux and use the widths on the next run.

The problem (as always) is compatibility: it's not clear this can easily be added to the standard classes without impacting the many packages that modify the table of contents layout.

It's certainly a feature that could be considered for any new classes though.

**Comment by dbitouze** [comment]
> You mean in case you spend 30 years enlarging `source2e.tex` and accidentally go over a thousand pages:-)

You read my mind! :wink:

> This has long been known as a common issue with larger documents (or documents with unusual page formatting) It wouldn't be particularly difficult to keep track of the width and store the maximum (or less computationally expensive just keep track of the largest page (and section) number and assume the typeset version is monotonically increasing in width) you could then write them to the aux and use the widths on the next run.

:+1: 

> The problem (as always) is compatibility: it's not clear this can easily be added to the standard classes without impacting the many packages that modify the table of contents layout.

Sigh...

> It's certainly a feature that could be considered for any new classes though.

Why not providing this as a by default not enabled feature, that could be enabled just with a single command?

**Comment by davidcarlisle** [comment]
> Why not providing this as a by default not enabled feature, that could be enabled just with a single command?

That is not impossible to consider, certainly more changes have been made in recent years than in earlier periods but it would mean adding `\if@wibble....\fi` in several places in the toc code, and the question is how many of

https://ctan.org/topic/toc-etc

patch at that point and would give errors or fail to work as intended if the patch failed....

That said, there need to be changes in that area to support pdf tagging, so if such a feature is going to be considered , adding it at the same time might be a possibility.

Note this isn't closed as not planned it's labeled as long term enhancement, whether that makes a difference in practice we'll see....

**Comment by FrankMittelbach** [comment]
But compatibility is not the only issue here. Just taking the maximum is in my experience seldom a good a idea, e.g. TLC3 would have looked rather bad this way and it had a bunch of such issues as well. For example, 10.10.10 which would have made the whole TOC look really ugly if everything would have been based on this. Or the LOT where we have 10.100 that has 5 digits plus a period while most entries have 2 or maximum of 3 digits. So basing everything on xx.xxx would be rather bad.

So basically what I'm saying is that yes there is something that should be offered but in many cases the simple maximum is not good enough for a quality result and insead you need to handtune the exceptions if they are few.

**Comment by claell** [comment]
I just independently ran into this problem by visually noticing numbers being not perfectly aligned vertically. For me it is much more subtle because I only have three-digit numbers and the issue also shows for that. I guess it is also slightly depending on the chosen font and font size.

But this is super problematic, I think. I would have expected this to not have happened. And really hope there are no other longstanding problems like that in LaTeX that I need to catch manually (by chance) and then fix locally. Because especially those subtle problems are pretty hard to catch optically (although the log will emit some warning, so it won't go silent).

Hopefully, there can be a proper fix integrated upstream soon, so one doesn't need manual fixes for such things.

Using LaTeX 3 syntax is not helping here, correct?
